File: Data_Structures/Stack/Golang/Stack.go
package stack

import (
	"fmt"

	"github.com/cheekybits/genny/generic"
)

// Item is a "generic type" our Stack will store
type Item generic.Type

// Stack is fundamentally just a slice of items with extra functionality.
type Stack []Item

// Push appends an item to the stack
func (s *Stack) Push(v Item) {
	*s = append(*s, v)
}

// Pop removes and returns the item at the end of the stack
func (s *Stack) Pop() Item {
	if s.IsEmpty() {
		fmt.Println(" The stack is already empty.")
		return ' '
	}
	v := (*s)[len(*s)-1]
	*s = (*s)[:len(*s)-1]
	return v
}

// Peek returns the Item at the end of the stack
func (s *Stack) Peek() Item {
	if s.IsEmpty() {
		fmt.Println(" The stack is already empty.")
		return ' '
	}
	return (*s)[len(*s)-1]
}

// IsEmpty returns a bool representing whether or not the stack is empty
func (s *Stack) IsEmpty() bool {
	return len(*s) == 0
}

File: Data_Structures/Stack/Java/Stack.java
package ds;
public class Stack {
	
	private int maxSize;
	private long[] stackArray;
	private int top;
	
	public Stack(int size) {
		this.maxSize = size;
		this.stackArray = new long[maxSize];
		this.top = -1;
	}
	
	public void push(long j) {
		if (isFull()) {
			System.out.println(" This stack is already full.");
		}
		else {
			top++;
			stackArray[top] = j;
		}
	}
	
	public long pop() {
		if (isEmpty()) {
			System.out.println(" This stack is already empty");
			return -1;
		}
		else {
			top--;
			return stackArray[top + 1];
		}
	}
	
	public long peek() {
		return stackArray[top];
	}
	
	public boolean isEmpty() {
		return (top == -1);
	}
	
	public boolean isFull() {
		return (maxSize-1 == top);
	}
	
}
